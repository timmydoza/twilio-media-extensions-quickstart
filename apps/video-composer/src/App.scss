@function get-vw($target) {
  $vw-context: (1280 / 100) * 1px;
  @return ($target / $vw-context) * 1vw;
}

$sidebarParticipants: 5;
$presentationBannerHeight: get-vw(44px);
$aspectRatio: 9/ 16;
$margin: get-vw(3px);

p {
  font-family: Inter, sans-serif;
  margin: 0;
  font-size: get-vw(16px)
}

video {
  width: 100%;
  height: 100%;
  display: flex;
  object-fit: cover;
}

.app {
  background: #121c2d;
  height: 100vh;
  display: flex;
  flex-direction: column;
  overflow: hidden;

  &.isPresentationModeActive {
    padding: $margin * 2;
    box-sizing: border-box;

    .mainContainer {
      display: grid;
      grid-template-rows: 1fr;
      // Here we calculate the width of the participant sidebar. The sidebar should be wide enough to hold exactly 5 participants with an aspect
      // ratio of 16:9 (accounting for margins), plus the indicator to show the "+ 1 more" indicator, which has the same height as the presentationBanner.
      grid-template-columns: 1fr calc(
          (100vh - #{$presentationBannerHeight} - (#{$margin} * 2)) / #{$sidebarParticipants} / #{$aspectRatio} - #{$margin}
        );
    }

    .participantContainer {
      position: relative;
      align-content: center;
      margin-top: -$margin;
      margin-right: -$margin * 2;
      margin-left: $margin;

      .participant:nth-child(n + #{$sidebarParticipants + 1}) {
        display: none;
      }
    }
  }
}

.presentationBanner {
  height: $presentationBannerHeight;
  margin-bottom: $margin * 2;
  display: none;
  align-items: center;
  justify-content: center;
  border-radius: get-vw(4px);
  background: #394762;
  color: white;
  display: flex;
}

.mainContainer {
  position: relative;
  flex: 1;
}

.presentationContainer {
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  border-radius: get-vw(4px);
  height: calc(100vh - #{$margin * 4});

  .presentationVideoContainer {
    display: flex;
    align-items: center;
    flex: 1;
    min-height: 0;

    div {
      width: 100%;
      max-height: 100%;
      border-radius: get-vw(4px);
      overflow: hidden;
    }

    video {
      max-height: calc(100vh - #{$presentationBannerHeight} - #{$margin * 8});
      object-fit: contain;
    }
  }
}

.participantContainer {
  position: absolute;
  display: flex;
  align-content: center;
  flex-wrap: wrap;
  justify-content: center;
  top: 0px;
  right: 0px;
  bottom: 0px;
  left: 0px;

  .participant:nth-child(n + 10) {
    display: none;
  }
}

.participantCountContainer {
  display: flex;
  align-items: center;
  justify-content: center;
  color: white;
  height: $presentationBannerHeight + $margin * 2;
}

.participant {
  position: relative;
  overflow: hidden;
  border-radius: get-vw(4px);

  p {
    position: absolute;
    bottom: get-vw(7.2px);
    left: get-vw(7.2px);
    border-radius: get-vw(4px);
    background: rgba(0, 0, 0, 0.5);
    padding: get-vw(2.88px) get-vw(4.32px);
    margin: 0;
    color: white;
    font-size: get-vw(14.4px);
    z-index: 1;
  }

  svg {
    position: absolute;
    top: get-vw(8px);
    right: get-vw(8px);
    z-index: 10;
  }
}

.noVideo {
  display: flex;
  align-items: center;
  justify-content: center;
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  z-index: 1;
  background-color: #8891aa;
  color: white;
  font-family: 'Inter', sans-serif;
  font-weight: 700;
  overflow-wrap: break-word;
  text-align: center;
}
