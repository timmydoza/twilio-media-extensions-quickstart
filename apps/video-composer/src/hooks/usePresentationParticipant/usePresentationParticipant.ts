import { useEffect, useState } from 'react';
import { RemoteParticipant, Room } from 'twilio-video';

/*
  Returns the participant that is presenting video content to the main view. This hook assumes that only one participant
  can present at a time.
*/

export default function usePresentationParticipant(room: Room | undefined) {
  const [presentationParticipant, setPresentationParticipant] = useState<RemoteParticipant>();

  useEffect(() => {
    if (room) {
      const updatePresentationParticipant = () => {
        setPresentationParticipant((prevPresentationParticipant) => {
          const participants = Array.from(room.participants.values());

          const presentationParticipants: (RemoteParticipant | undefined)[] = participants.filter((participant) => {
            const participantTrackPublications = Array.from(participant.tracks.values());
            return participantTrackPublications.some(
              (trackPublication) =>
                trackPublication.trackName.includes('video-composer-presentation') && trackPublication.kind === 'video'
            );
          });

          // If the active presentationParticipant (the 'prevPresentationParticipant') is not in the list of current presentationParticipants,
          // then replace them with the first presentationParticipant in the list (which could be undefined).

          // This prevents a new presentationParticipant from being the active presentationParticipant when there
          // is already an active presentationParticipant.
          if (!presentationParticipants.includes(prevPresentationParticipant)) {
            return presentationParticipants[0];
          } else {
            return prevPresentationParticipant;
          }
        });
      };

      updatePresentationParticipant();

      room.on('trackPublished', updatePresentationParticipant);
      room.on('trackUnpublished', updatePresentationParticipant);
      room.on('participantConnected', updatePresentationParticipant);
      room.on('participantDisconnected', updatePresentationParticipant);

      return () => {
        room.off('trackPublished', updatePresentationParticipant);
        room.off('trackUnpublished', updatePresentationParticipant);
        room.off('participantConnected', updatePresentationParticipant);
        room.off('participantDisconnected', updatePresentationParticipant);
      };
    }
  }, [room]);

  return presentationParticipant;
}
