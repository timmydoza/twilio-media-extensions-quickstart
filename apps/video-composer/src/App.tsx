import './App.scss';
import 'normalize.css';

import { ASPECT_RATIO, SIDEBAR_PARTICIPANTS } from './constants';
import clsx from 'clsx';
import { useAutoDisconnect, useRoom } from '@media-extensions/commons';
import Participant from './components/Participant/Participant';
import usePresentationParticipant from './hooks/usePresentationParticipant/usePresentationParticipant';
import PresentationParticipant from './components/PresentationParticipant/PresentationParticipant';
import useParticipants from './hooks/useRemoteParticipants/useRemoteParticipants';
import useDominantSpeaker from './hooks/useDominantSpeaker/useDominantSpeaker';
import useTimedMetadata from './hooks/useTimedMetadata/useTimedMetadata';
import useGridLayout from './hooks/useGridLayout/useGridLayout';

export default function App() {
  const room = useRoom({ isVideo: true });
  const participants = useParticipants(room);
  useAutoDisconnect(room, participants);
  useTimedMetadata(room, participants);
  const dominantSpeaker = useDominantSpeaker(room);
  const presentationParticipant = usePresentationParticipant(room);
  const isPresentationModeActive = Boolean(presentationParticipant);

  const MAX_PARTICIPANTS = isPresentationModeActive ? SIDEBAR_PARTICIPANTS : 9;

  const { participantVideoWidth, containerRef } = useGridLayout(
    Math.min(participants.length, MAX_PARTICIPANTS),
    isPresentationModeActive
  );

  const participantsArray = isPresentationModeActive
    ? [presentationParticipant, ...participants.filter((p) => p !== presentationParticipant)] // Put presentation participant at the front of the list
    : participants;

  const participantWidth = `${participantVideoWidth}px`;
  const participantHeight = `${Math.floor(participantVideoWidth * ASPECT_RATIO)}px`;

  return (
    <div
      className={clsx('app', {
        isPresentationModeActive,
      })}
    >
      <div className="mainContainer">
        {isPresentationModeActive && (
          <div className="presentationContainer">
            <div className="presentationBanner">
              <p data-cy-presentation-banner>{presentationParticipant!.identity} is presenting.</p>
            </div>
            <div className="presentationVideoContainer" data-cy-presentation-container>
              <PresentationParticipant participant={presentationParticipant!} />
            </div>
          </div>
        )}
        <div ref={containerRef} className="participantContainer">
          {participantsArray.map((participant) => (
            <Participant
              key={participant!.identity}
              participant={participant!}
              width={participantWidth}
              height={participantHeight}
              isDominantSpeaker={dominantSpeaker === participant}
              fontSize={participantVideoWidth / 18 + 'px'}
            />
          ))}
          {participants.length > SIDEBAR_PARTICIPANTS && isPresentationModeActive && (
            <div className="participantCountContainer">
              <p>+ {participants.length - SIDEBAR_PARTICIPANTS} more</p>
            </div>
          )}
        </div>
      </div>
    </div>
  );
}
