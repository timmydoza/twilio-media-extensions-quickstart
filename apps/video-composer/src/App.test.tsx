import App from './App';
import { shallow } from 'enzyme';
import usePresentationParticipant from './hooks/usePresentationParticipant/usePresentationParticipant';

const mockParticipants = [
  { identity: 'test-participant-1' },
  { identity: 'test-participant-2' },
  { identity: 'test-participant-3' },
];

jest.mock('@media-extensions/commons', () => ({
  ...jest.requireActual('@media-extensions/commons'),
  useRoom: jest.fn(),
}));
jest.mock('./constants', () => ({ SIDEBAR_PARTICIPANTS: 2, ASPECT_RATIO: 9 / 16 }));
jest.mock('./hooks/useRemoteParticipants/useRemoteParticipants', () => () => mockParticipants);
jest.mock('./hooks/useDominantSpeaker/useDominantSpeaker', () => () => mockParticipants[0]);
jest.mock('./hooks/usePresentationParticipant/usePresentationParticipant');
jest.mock('./hooks/useGridLayout/useGridLayout', () => () => ({
  participantVideoWidth: 720,
  containerRef: 'mockRef',
}));

const mockUsePresentationParticipant = usePresentationParticipant as jest.Mock<any>;

describe('App component', () => {
  it('should render correctly', () => {
    const wrapper = shallow(<App />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly when presentation mode is active', () => {
    mockUsePresentationParticipant.mockImplementationOnce(() => mockParticipants[1]);
    const wrapper = shallow(<App />);
    expect(wrapper).toMatchSnapshot();
  });
});
