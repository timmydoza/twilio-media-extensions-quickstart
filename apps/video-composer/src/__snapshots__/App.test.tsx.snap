// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component should render correctly 1`] = `
<div
  className="app"
>
  <div
    className="mainContainer"
  >
    <div
      className="participantContainer"
    >
      <Participant
        fontSize="40px"
        height="405px"
        isDominantSpeaker={true}
        key="test-participant-1"
        participant={
          Object {
            "identity": "test-participant-1",
          }
        }
        width="720px"
      />
      <Participant
        fontSize="40px"
        height="405px"
        isDominantSpeaker={false}
        key="test-participant-2"
        participant={
          Object {
            "identity": "test-participant-2",
          }
        }
        width="720px"
      />
      <Participant
        fontSize="40px"
        height="405px"
        isDominantSpeaker={false}
        key="test-participant-3"
        participant={
          Object {
            "identity": "test-participant-3",
          }
        }
        width="720px"
      />
    </div>
  </div>
</div>
`;

exports[`App component should render correctly when presentation mode is active 1`] = `
<div
  className="app isPresentationModeActive"
>
  <div
    className="mainContainer"
  >
    <div
      className="presentationContainer"
    >
      <div
        className="presentationBanner"
      >
        <p
          data-cy-presentation-banner={true}
        >
          test-participant-2
           is presenting.
        </p>
      </div>
      <div
        className="presentationVideoContainer"
        data-cy-presentation-container={true}
      >
        <PresentationParticipant
          participant={
            Object {
              "identity": "test-participant-2",
            }
          }
        />
      </div>
    </div>
    <div
      className="participantContainer"
    >
      <Participant
        fontSize="40px"
        height="405px"
        isDominantSpeaker={false}
        key="test-participant-2"
        participant={
          Object {
            "identity": "test-participant-2",
          }
        }
        width="720px"
      />
      <Participant
        fontSize="40px"
        height="405px"
        isDominantSpeaker={true}
        key="test-participant-1"
        participant={
          Object {
            "identity": "test-participant-1",
          }
        }
        width="720px"
      />
      <Participant
        fontSize="40px"
        height="405px"
        isDominantSpeaker={false}
        key="test-participant-3"
        participant={
          Object {
            "identity": "test-participant-3",
          }
        }
        width="720px"
      />
      <div
        className="participantCountContainer"
      >
        <p>
          + 
          1
           more
        </p>
      </div>
    </div>
  </div>
</div>
`;
